### PR TITLE
Support for multiple arrays in map_overlap

### PR DESCRIPTION
**Pull request by eric-czech**
- [X] Tests added / passed
- [X] Passes `black dask` / `flake8 dask`

This is an attempt to address https://github.com/dask/dask/issues/2403.

A summary of the changes here:

- The signature of `map_overlap` was changed to `map_overlap(func, *args, depth=None, ...)` to better match `map_blocks(func, *args, ...)`
  - My changes include detecting use of the old signature, throwing a warning, and converting the arguments where necessary (per @mrocklin's suggestion)
  - `da.overlap.map_overlap` was only called previously from `Array.map_overlap` (so it was easy to switch to the new signature), but this check should avoid issues for any users calling the module method directly
- All arrays can be of differing ranks but they must broadcast (i.e. they cannot have different sizes in any one dimension, if not unit size).  The broadcasting is not actually performed to avoid copying small arrays to match the size of larger ones.  Users could easily perform this broadcasting first though if that's the behavior they want.
- Chunk sizes can be different and are aligned where necessary by default, but users can opt out of this behavior using `align_arrays=False` (much like in `blockwise`).  I think this is an important flag because at least personally, I wouldn't want the automatic alignment creating non-uniformly sized chunks or very small chunks in large datasets where the block sizes were chosen carefully.
- Depth now has a default of 0 rather than being a required positional argument like before.  This means that given nothing other than a function and some arrays, `map_overlap` is a pass-through to `map_blocks`.
- `boundary` and `depth` arguments are interpreted as a sequence of arguments corresponding to each array if given as `list` type.
- The final trim step is done using `depth` and `boundary` for the array of largest rank since this seemed to be the most intuitive behavior, and is reverse compatible. 

I didn't add any examples to the docs yet, but I wanted to submit this for a preliminary look first before trying to put those finishing touches on it.  Let me know if there are any major issues here.

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @jakirkham if you have time

File: dask/array/overlap.py [hunk 1]
```diff
@@ -584,29 +608,83 @@ def map_overlap(x, func, depth, boundary=None, trim=True, **kwargs):
...
+    # Look for invocation using deprecated single-array signature
+    # map_overlap(x, func, depth, boundary=None, trim=True, **kwargs)
+    if isinstance(func, Array) and callable(args[0]):
+        warnings.warn(
+            "Detected use of signature map_overlap(x, func) rather than "
+            "map_overlap(func, *args) for multi-array support. Arguments "
+            "will be swapped in this case but such an exception will not "
+            "be made in a future release.",
+            FutureWarning,
+        )
```
[the remark this answers is not in the document]
Comment (jni) [reply dask/array/overlap.py:620]: I don't know of any, and I like this update! :crossed_fingers: :grimacing: 

I would maybe rephrase the message to be more explicit about how to fix this:

The use of `map_overlap(array, function, **kwargs)` is deprecated since dask x.y and will be an error in dask z.w. To silence this message, use the syntax `map_overlap(function, array0,[ array1, ...,] **kwargs)`.

**Comment by eric-czech** [comment]
Thanks for the feedback guys!  I'll make those changes, add some examples to the docs, and commit those updates soon(ish).

File: dask/array/overlap.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eric-czech) [reply dask/array/overlap.py:620]: Thanks @jni, I'll make that change.

[the remark this answers is not in the document]
Comment (eric-czech) [reply dask/array/overlap.py:659]: I updated the message assuming that this change would be pushed in 2.17.  Not sure if that will be true, and I also wasn't sure when to say it would be an error so I hedged by saying it would happen in some future release.  This message seem reasonable now?

**Comment by eric-czech** [comment]
The last two commits contain @mrocklin's small changes in the tests,  @jni's change to the signature warning, and some more descriptions/examples that I added to the docs.  Let me know if there's anything else that should be updated.

**Comment by chrisroat** [comment]
This is really awesome.  I think it would be useful to add a test to tease the codepath with depth as a list.

**Comment by eric-czech** [comment]
Welp, I added a test for that @chrisroat but these CI errors are killing me.  Does anyone know off-hand what I'm doing incorrectly [here](https://travis-ci.org/github/dask/dask/jobs/690055990):

```
> da.map_overlap(func, x, y, depth=1).compute() # doctest: +NORMALIZE_WHITESPACE
Expected:
    array([[0,  2,  4,  6],
           [4,  6,  8,  10]])
Got:
    array([[ 0,  2,  4,  6],
           [ 4,  6,  8, 10]])
```

With or without `doctest: +NORMALIZE_WHITESPACE` or with/without tabs/spaces in the doc string I get this error (in any combination).  The source code looks like this:

```
    """
    >>> func = lambda x, y: x + y
    >>> x = da.arange(8).reshape(2, 4).rechunk((1, 2))
    >>> y = da.arange(4).rechunk(2)
    >>> da.map_overlap(func, x, y, depth=1).compute() # doctest: +NORMALIZE_WHITESPACE
    array([[0,  2,  4,  6],
           [4,  6,  8,  10]])
    """
```

**Comment by eric-czech** [comment]
fyi I don't get that doctest error locally in a Python 3.8.2 environment (i.e. w/ `py.test dask --doctest-modules`), yet 3.8 is the only one failing during the build (3.6 and 3.7 are fine).

**Comment by jsignell** [comment]
I can reproduce the doctest failures locally, so I'll try to see if I can come up with a suggestion.

File: dask/array/overlap.py [hunk 2]
```diff
@@ -583,30 +609,121 @@ def map_overlap(x, func, depth, boundary=None, trim=True, **kwargs):
...
-                "for boundary='none', however boundary for dimension "
-                "{} is {}".format(i, boundary2[i])
+    The ``da.map_overlap`` function can also accept multiple arrays.
+
+    >>> func = lambda x, y: x + y
+    >>> x = da.arange(8).reshape(2, 4).rechunk((1, 2))
+    >>> y = da.arange(4).rechunk(2)
+    >>> da.map_overlap(func, x, y, depth=1).compute() # doctest: +NORMALIZE_WHITESPACE
+    array([[0,  2,  4,  6],
+           [4,  6,  8,  10]])
```
Comment (gforsyth) [inline dask/array/overlap.py:620]:
```suggestion
    array([[ 0,  2,  4,  6],
           [ 4,  6,  8,  10]])
```

numpy likes to pad out the array reprs and `NORMALIZE_WHITESPACE` treats blocks of whitespace the same way, but `[0` and `[ 0` are viewed as different things.  You can try the above to see if it resolves the doctest failures.

**Comment by eric-czech** [comment]
Thanks @jsignell/@gforsyth!  I committed Gil's changes so hopefully that clears.

**Comment by jsignell** [comment]
Oh man! I didn't realize I was in a race with Gil :) 

I found one more thing to change:

```diff
diff --git a/dask/array/overlap.py b/dask/array/overlap.py
index 33be1ffb..0bd54ea9 100644
--- a/dask/array/overlap.py
+++ b/dask/array/overlap.py
@@ -631,8 +631,10 @@ def map_overlap(
     >>> len(r.to_delayed())
     4
 
-    >>> da.map_overlap(func, x, y, depth=1, align_arrays=False)
-    ValueError: Shapes do not align
+    >>> da.map_overlap(func, x, y, depth=1, align_arrays=False).compute()
+    Traceback (most recent call last):
+        ...
+    ValueError: Shapes do not align {'.0': {2, 4}}
 
     Note also that this function is equivalent to ``map_blocks``
     by default.  A non-zero ``depth`` must be defined for any
```

**Comment by jsignell** [comment]
I know that `Traceback` part looks wrong, but apparently it doesn't matter what you put in it https://docs.python.org/3.8/library/doctest.html#what-about-exceptions

File: dask/array/overlap.py [hunk 3]
```diff
@@ -583,30 +609,121 @@ def map_overlap(x, func, depth, boundary=None, trim=True, **kwargs):
...
+    so the blocks must have a common chunk size.  This common chunking
+    is determined automatically as long as ``align_arrays`` is True.
+
+    >>> x = da.arange(8, chunks=4)
+    >>> y = da.arange(8, chunks=2)
+    >>> r = da.map_overlap(func, x, y, depth=1, align_arrays=True)
+    >>> len(r.to_delayed())
+    4
+
+    >>> da.map_overlap(func, x, y, depth=1, align_arrays=False)
```
Comment (jsignell) [inline dask/array/overlap.py:634]:
```suggestion
    >>> da.map_overlap(func, x, y, depth=1, align_arrays=False).compute()
    Traceback (most recent call last):
        ...
```

File: dask/array/overlap.py [hunk 4]
```diff
@@ -583,30 +609,121 @@ def map_overlap(x, func, depth, boundary=None, trim=True, **kwargs):
...
+    ValueError: Shapes do not align
```
Comment (jsignell) [inline dask/array/overlap.py:635]:
```suggestion
    ValueError: Shapes do not align  {'.0': {2, 4}}
```

File: dask/array/overlap.py (hunk 3 above)
Comment (jsignell) [reply dask/array/overlap.py:634]: In case this is easier than applying the diff :)

**Comment by eric-czech** [comment]
Thanks @jsignell, committed and building again.

**Comment by jsignell** [comment]
@dask/maintenance tests are now passing if someone wants to take another look and merge this in. Thanks for being so patient @eric-czech this looks like a great improvement!

**Comment by eric-czech** [comment]
Sure thing, thanks again for the help getting it across the finish line.

**Comment by jakirkham** [comment]
Thanks for the work on this @eric-czech and everyone for the reviews! 😄
